### PR TITLE
Ignore PhantomData when checking CoerceUnsized implementations

### LINKED ISSUE #26905: DST coercions are prevented by phantom data
This prevents Rc from having a sound definition with respect to dropcheck

```rust
#![feature(unsize, coerce_unsized)]

use std::marker::{Unsize, PhantomData};
use std::ops::CoerceUnsized;

struct MyRc<T: ?Sized> {
    ptr: *const T,
    _boo: PhantomData<T>,
}

impl<T: ?Sized + Unsize<U>, U: ?Sized> CoerceUnsized<MyRc<U>> for MyRc<T>{ }

fn main() {}
```

```text
<anon>:11:1: 11:77 error: the trait `CoerceUnsized` may only be implemented for a coercion between structures with one field being coerced, but 2 fields need coercions: ptr (*const T to *const U), _boo (core::marker::PhantomData<T> to core::marker::PhantomData<U>) [E0375]
<anon>:11 impl<T: ?Sized + Unsize<U>, U: ?Sized> CoerceUnsized<MyRc<U>> for MyRc<T>{ }
          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
error: aborting due to previous error
```

### PR DESCRIPTION
**Pull request by Gankra**
Fixes #26905 

May want to wait on https://github.com/rust-lang/rfcs/pull/1234

I'm not super happy with the duplicated logic in trans and typeck, and would love to know if there's a "better way". It's pretty minimal, regardless.

r? @nrc

### CODE REVIEW

File: src/librustc_trans/trans/expr.rs [hunk 1]
```diff
@@ -537,8 +537,15 @@ fn coerce_unsized<'blk, 'tcx>(bcx: Block<'blk, 'tcx>,
...
                 } else {
+                    let is_phantom = if let &ty::TyStruct(def_id, _) = &src_ty.sty {
+                        Some(def_id) == bcx.tcx().lang_items.phantom_data()
+                    } else {
+                        false
+                    };
                     // Otherwise, simply copy the data from the source.
-                    assert_eq!(src_ty, target_ty);
+                    if !is_phantom {
+                        assert_eq!(src_ty, target_ty);
```
Comment (nrc) [inline src/librustc_trans/trans/expr.rs]: assert!(is_phantom || arc_ty == target_ty);

File: src/librustc_typeck/coherence/mod.rs [hunk 2]
```diff
@@ -485,12 +485,25 @@ impl<'a, 'tcx> CoherenceChecker<'a, 'tcx> {
...
                     let fields = tcx.lookup_struct_fields(def_id_a);
+                    let phantom_data = tcx.lang_items.phantom_data();
+
                     let diff_fields = fields.iter().enumerate().filter_map(|(i, f)| {
                         let ty = tcx.lookup_field_type_unsubstituted(def_id_a, f.id);
                         let (a, b) = (ty.subst(tcx, substs_a), ty.subst(tcx, substs_b));
+
+                        // Ignore PhantomData -- doesn't matter
+                        if let &ty::TyStruct(def_id, _) = &ty.sty {
+                            if Some(def_id) == phantom_data {
```
Comment (nrc) [inline src/librustc_typeck/coherence/mod.rs]: may as well inline phantom_data

File: src/test/run-pass/issue-26905.rs [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+}
+
+impl<T: ?Sized + Unsize<U>, U: ?Sized> CoerceUnsized<MyRc<U>> for MyRc<T>{ }
+
+fn main() {
+    let data = [1, 2, 3];
+    let iter = data.iter();
+    let x = MyRc { _ptr: &iter, _boo: PhantomData };
+    let _y: MyRc<Iterator<Item=&u32>> = x;
+}
```
Comment (nrc) [inline src/test/run-pass/issue-26905.rs]: Should add a test that other ZSTs still cause errors.

File: src/librustc_typeck/coherence/mod.rs (hunk 2 above)
Comment (Gankro) [inline src/librustc_typeck/coherence/mod.rs]: Wasn't sure if that was a perf issue (e.g. I've seen it cached elsewhere).

**Comment by nrc** [comment]
r=me with an extra test and nits addressed, should wait for the RFC before landing though

**Comment by Gankro** [comment]
nits addressed, I think.

**Comment by nrc** [comment]
Thanks! r=me, but waiting on RFC

**Comment by eddyb** [comment]
LGTM. You could deduplicate the check by adding an `is_phantom` method to `middle::ty::Ty`.

**Comment by ebfull** [comment]
Would this implementation allow me to do this?

```rust
#![feature(coerce_unsized, unsize)]

use std::marker::{PhantomData, Unsize};
use std::ops::CoerceUnsized;

trait Bar { }

struct Foo<T: Bar + ?Sized> {
    a: PhantomData<T>
}

impl<T: Bar + ?Sized + Unsize<U>, U: Bar + ?Sized> CoerceUnsized<Foo<U>> for Foo<T> { }

fn main() { }
```

I have an API which uses phantom types in this way, and right now `CoerceUnsized` can't be implemented for my type because there is no blanket `CoerceUnsized` impl for `PhantomData`.

**Comment by Gankro** [comment]
superseded

**Pull request closed without merging**